Switch to Ubuntu 12.04/Precise for PHP 5.3

cjhaas opened the pull request:
Per the [blog post](https://blog.travis-ci.com/2017-08-31-trusty-as-default-status), [migration guide](https://docs.travis-ci.com/user/precise-to-trusty-migration-guide/) and [this issue](https://github.com/travis-ci/travis-ci/issues/8219) we need to explicitly state `dist: precise` to get Ubuntu 12.04/Precise in order to support PHP 5.3

cjhaas commented:
Looks like the choice for Travis is either supporting PHP 5.3 or HHVM, you can't have both.

VasekPurchart commented:
Lately from what I have seen many repos are [dropping support](https://symfony.com/blog/symfony-4-end-of-hhvm-support) for HHVM, also considering what they [announced](http://hhvm.com/blog/2017/09/18/the-future-of-hhvm.html) (departure from PHP compatibility) I think it's adequate. Also there were always problems with HHVM on Twitter regarding which version (and compatible with which PHP version) is actually used.

Btw I am also not sure whether it is necessary to support PHP 5.3 in new releases since it does not even receive security fixes (bar features) for 3 years. If someone wants to use this package with so old PHP, then there are plenty of versions already released.

cjhaas commented:
@VasekPurchart, I'm for giving up on HHVM since they're going in a different direction and everyone else seems to be dropping active support.

I have a specific project that needs legacy PHP support still which is why I'm using your code. I say keep PHP 5.3 around until it becomes a burden to maintain your code and then just create a new release for modern PHP. Everyone can just ` || ` in their composer file then if they really need the legacy version.

The third changeset in this PR puts you back on Ubuntu 12.04/Precise which is still actively supported by Travis and still runs in the Docker environment (at least, I'm 99% sure it does) since you have `sudo: false`.

I can add one last commit if you want to drop HHVM

VasekPurchart commented:
Just to avoid confusion I am not an official maintainer of this package, just occasional contributor, so the decision is not up to me, was just giving my opinion based on the trends I follow (and what I do in my opensource packages).

cjhaas commented:
Thanks @VasekPurchart, I understand!

I added a fourth commit that switches Travis over to a matrix in order to flag HHVM as a failure for now since that's what [PHP Code Sniffer does](https://github.com/squizlabs/PHP_CodeSniffer/blob/master/.travis.yml). I also added PHP 7.2 and nightly into the mix for the same reason.

JakubOnderka commented:
Hi, @cjhaas! I just switched to `dist: precise` and removed support for HHVM, because probably nobody use it. And because I don't want to make big changes in this project, I want to keep support even for old PHP version.

But this pull request cannot be merged, because contains changes from other branches. So I put changes directly to master.

Thanks @cjhaas and @VasekPurchart
